### DISCUSSION TITLE
Contributor call: 20 August 2022

### DISCUSSION THREAD

**Discussion by joostdecock** [freesewing/freesewing#2582] [Support] [answered]
You are all invited to the FreeSewing Contributor Call.
Please leave a comment here or [on Discord](https://discord.freesewing.org/) to add things to the agenda.
This issue will be updated after the call with notes and to-dos.

### About
FreeSewing's contributor call is an informal hangout/chat/meeting that is open to all FreeSewing contributors and users alike.
The call is organized/hosted by @tangerineshark.
The call does not have a fixed structure or agenda items. Everyone is free to suggest agenda items, ask questions, or bring ideas.

Details:
Location: https://discord.freesewing.org (meeting room: 🔊voice chat)
Date: 20 August 2022 4pm UTC, which can be found in your local time zone [here](https://savvytime.com/converter/utc-to-belgium-antwerpen-ca-los-angeles-il-chicago-ny-new-york-city-brazil-sao-paulo-united-kingdom-london-france-paris-south-africa-johannesburg-msd-india-mumbai-hkt-japan-tokyo-australia-sydney/aug-20-2022/4pm).
Or check the bot in Discord! #meetings
The meetings will always be at 6pm, Antwerp time.

### Agenda
- Anyone willing to take notes?
- The ongoing discussion/work on adding parts
  - See:
    - #2538 
    - #1107 
    - #2546 
  - Whether or not these changes should bump the major version
  - What changes would/could be breaking?
  - Does anybody plan to implement a different proposal from #2546
  - What is lacking from the current #2546 
- Who do you want to hear from in the next newsletter? And what else would you like to read?

Notes will be added after the call

**Comment by joostdecock** [comment]
In preparation of the call, here's an attempt to summarize the current situation regarding #2546 
As such, it is also a summary of the current status of #2538 and #1107

## How to export your parts
Essentially, the following configuration is now managed at the part level:

- `options`
- `measurements`
- `optionalMeasurements`
- `dependencies`
- `inject`
- `hide`
 
To facilitate that, a part should now export a pain object (aka POJO for plain old javascript object) with the following properties:

| Name | Mandatory | Description |
| ------ | -- | -- |
| `name` | **Yes** | Each part needs a name. This is it. To avoid naming clashes in inheritance, there will be a new naming convention like this: `design:part`. <br />So for example `brian:front` or `simon:sleeve`  |
| `draft` | **Yes** | The actual draft method (what we used to attach tot he pattern prototype) |
| `after` | No | A simple part dependency. As in, draft this after that (what used to be configured under `dependencies`. An exported part object, or an array of them. |
| `inject` | No | A extending part dependency. As in, draft this starting from that (what used to be configured under `inject`. An exported part object. |
| `hide` | No | Set this to `true` if you want to hide the part buy default (what used to be configured under `hide`. |
| `options` | No | An object of option names / option config pairs. Like what used to be configured under `options`. |
| `measurements` | No | An array of measurement names. Like what used to be configured under `measurements`. |
| `optionalMeasurements` | No | An array of optional measurement names. Like what used to be configured under `optionalMeasurements`. |

The only thing that is left out to make things *just work* are the `optionGroups`. Which brings me to:

## How to construct your design

The `freesewing.Design()` constructor now takes a single POJO as input with the following properties:

| Name | Description |
| ------ | -- |
| `parts` | An array of exported part objects  |
| `plugins` | An array of plugins. (Like the second parameter to the `Design()` constructor today)  |
| `conditionalPlugins` | An array of conditional plugins. (Like the third parameter to the `Design()` constructor today)  |

**All other attributes** will be added to `pattern.config`. If the attribute exists, we will do a **shallow merge**.
> This is how we'll handle `optionGroups` (see below)

## How to export your design

You should export each part as a named export, and the design with a capital as it's a constructor.
The design itself should also be the default export:

```js
const Aaron = new freesewing.Design({
  ...info,
  measurements,
  optionalMeasurements,
  options: { ...options },
  parts: { back, front },
  plugins: pluginBundle,
  conditionalPlugins: {
    plugin: bustPlugin,
    condition: (settings=false) =>
      settings?.options?.draftForHighBust &&
      settings?.measurements?.highBust
      ? true : false
  }
})

// Named exports
export { front, back, Aaron }
// Default export
export default Aaron
```

## `optionGroups` and other custom configuration

The one thing missing to make all of this work like magic in the UI is `optionGroups`.
Problem is, optionGroups is not part of the FreeSewing core API. It is just something that we use to structure how users can interact with the options in the frontend.

So I don't want to add something *specific* for `optionGroups` to core because it's a unwieldy complicated object with a mixture of arrays and POJOs to indicate hierarchical group structure.

Instead, we'll fold under the general rule that *anything we don't recognize as a core-specific attribute will be merged into the config*.

So, `optionGroups` will just be another thing that core doesn't recognize, and thus merges into the config.
To make that work, we'll need to re-structure how we define option groups because the current structure does not allow shallow merging.

So instead of this:
```js
{
  fit: [ 'chestEase', 'collarEase' ],
  advanced: [
  'acrossBackFactor',
  'shoulderSlopeReduction',
  {
    sleevecap: [ 'sleevecapEase' ],
  }
}    
```

We'll use this structure:

```js
{
  fit: [ 'chestEase', 'collarEase' ],
  advanced: [ 'acrossBackFactor', 'shoulderSlopeReduction' ],
  "advanced:sleevecap": [ 'sleevecapEase' ],
}    
```
Essentially, a flat structure for core to merge, but the use of the `:` separator allows us to reconstruct the hierarchy in our UI code.

## v2 or v3?

Over the last week or so, I've implemented most of this, and a lot of it I was able to do without any breaking changes.
Or at least, breaking changes in the core library.

But there's a few issues that are cause for concern:

### Named exports are great, but don't work in CJS

The new way of handling parts makes it much easier to extend another pattern, as pulling in the part will automatically come with all the config that part needs. It will also follow the dependency chain, so if you want `brian:sleeve` you don't have to know it depends on `brian:sleevecap` it will *just work.

To make that work, we export the parts as named exports. But named exports don't exist in CommonJS (CJS) modules. Only the default export is available. Nothing that you can't work around, but it would be a breaking change for those peope who use CJS.

Who uses CJS? Good question, since we ship both ESM and CJS most people don't even know/realize what they are using under the hood.

### Config is only available after pattern initialization

This is also not super documented, but currently the (default export of) the pattern has a `.config` property that holds the pattern configuration. That was added so you could get the pattern config (and for example figure out what measurements it needs) without having to instantiate the pattern.

However, as part of the work, we now have support for adding parts at run time. This late-stage adding of patterns means that generating the full config (including the dependency chain) is deferred to the draft stage, and so the config object is no longer available without instantiating the pattern.

This, for example, breaks our very own `pattern-info` package since it relies on that static config.

### Backwards compatibility is possible, but not great

While adding this with core-level backwards compatibility was possible, it doesn't really allow for clean code.
Things would be much easier if we could double-down on these new features without having to drag the rest along.

## Conclusion

What's been written now enabled one of the more far-fetched ideas on our roadmap: *Support for packs*.
This has now become trivial as every pattern can be a pack, it's merely a matter of defining the requirements to use that pack. Which I'm not entirely certain is a code thing, it might just be a documentation thing.

One of the hardest things to deal with is the constant tension between CJS and ESM and with a lot of these new features relying on named exports, CJS users are going to end up with breaking changes anyway.

Even if we can handle this without breaking changes for all our designs. Things will break for people who extend our current designs **and** rely on the staticly exported config to re-use some of the config.
This is actually something that's currently breaking some or our own designs.

This will require a bunch of documentation updates and it would be difficult to always explain both versions next to each other.

So, here's what I propose:

- We release 2.22 with Wouter's new Octoplushy design
- We fork that version of the `develop` branch into a new `v2` branch
- The `develop` branch will from now on be dedicated to the development of `v3` but it will be published as beta only
- freesewing.org will remain on v2 until we are ready with the new website
- We will setup v2.freesewing.dev with a snapshot of the latest v2 docs
- freesewing.dev will continue to carry the latest docs, which would be v3 docs. There will be a message that v2 docs are elsewhere
- The lab will run v3 code only
- `@freesewing/new-design` will also work with v3 code.- We will inform our userbase in the newsletter/blog post and so on that we've officially started working on v3

Essentially, everything moves to v3 except our users :)

**Comment by BenJamesBen** [comment]
I am fine with freesewing.org being frozen at the v2 snapshot to allow progress to continue elsewhere. Some things came to mind:

- It should be made easy to tell what version of the website you are on. Perhaps something that says "v2.21.1" on the page, for example.
- Possibly, it should be made difficult for customers to accidentally switch from a v2 to v3 site. I am thinking of the links to freesewing.dev that are currently found on freesewing.org web pages. I am also thinking of links to GitHub found on freesewing.org (since GitHub has links back to freesewing.dev).
- The idea is to avoid a situation where customers are confused about exactly where they are and what things are (and aren't) available to them there. Or, situations where a customer ends up on the 'wrong" site.
- There should be a place to see an overview of what versions all the different sites are running, perhaps with a list/summary of the differences/changes.
- If we fix a bug (in a design, for example) or add new documentation, what is the expectation for when it will become available on freesewing.org? Are customers stuck with all the old, possibly incorrect stuff for the next X years?
- Similarly, when a new design comes out, when will it become available on freesewing.org?
- Or, will high-priority bug fixes and new content be backported to v2 while v3 development is ongoing? If so, what is the process for this? (Determining what will be backported. Keeping track of the changes/differences in each version.)
- What is the estimated timeframe for expecting freesewing.org to move to v3?

**Reply by joostdecock** [reply]
Let me try and see if I can address some of these questions:

> It should be made easy to tell what version of the website you are on. Perhaps something that says "v2.21.1" on the page, for example.

Any v3 site will clearly be labeled as being beta, in addition to being hosted on next.freesewing.org or perhaps we'll change that at some point to v3.freesewung.org

We could add the version on the current site, but I'm not sure it matters so much. The version is always included on the pattern. We had the version in the footer at one point. When we removed it, nobody ever complained.

> Possibly, it should be made difficult for customers to accidentally switch from a v2 to v3 site. I am thinking of the links to freesewing.dev that are currently found on freesewing.org web pages. I am also thinking of links to GitHub found on freesewing.org (since GitHub has links back to freesewing.dev).

Links go to domains. When we link to FreeSewing.org that will link to the production site, whatever version it runs.

> The idea is to avoid a situation where customers are confused about exactly where they are and what things are (and aren't) available to them there. Or, situations where a customer ends up on the 'wrong" site.

That seems like it would be rather unusual, but ok.

> There should be a place to see an overview of what versions all the different sites are running, perhaps with a list/summary of the differences/changes.

We can create an overview of different sites, but keeping version numbers in sync, let alone a list of changes sounds like a lot of overhead for little benefit.

Also keep in mind that the current site is tightly coupled with released versions.
The new site is not. We could never ever release a new version again and still have new designs appear in the site because the entire site is built from source rather than packaged dependencies.

The packeges we ship will be entirely for the benefit of other developers.

> If we fix a bug (in a design, for example) or add new documentation, what is the expectation for when it will become available on freesewing.org? Are customers stuck with all the old, possibly incorrect stuff for the next X years?
Similarly, when a new design comes out, when will it become available on freesewing.org?
Or, will high-priority bug fixes and new content be backported to v2 while v3 development is ongoing? If so, what is the process for this? (Determining what will be backported. Keeping track of the changes/differences in each version.)

If it's important enough, it will be backported.
How? By me unless somebody else does it 🙂
V2 will live in it's own branch with it's own changelog. I don't think that will be an issue.

> What is the estimated timeframe for expecting freesewing.org to move to v3?

V3 will go live in 2023 😬

**Comment by KaerMorhan** [comment, marked as answer]
**NOTES**

Participants:

- @joostdecock 
- @KaerMorhan
- @bobgeorgethe3rd 
- @tangerineshark 
- @lexander2002
- @bbdapalkow
- @OysteinHoiby 
- @eleonoremilbou
- @SeaZeeZee
- @eriese 
- @nataliasayang
- illagerBill9
- Benjamin F.
- Nchilada
- joe_totale
- Simon J

✨ **Joost**

- FreeSewing v3

There is a roadmap that marks what can be minimally changed without breaking the website, what needs to be redone and done completely from scratch to plan out the progression of the website.
Joost and Enoch are working on new layout of the patterns that would allow people to change the layout themselves, as well as separately show the layout on paper and for cutting. It is currently hard to do with existing core library so Joost and Enoch are working on the code for adding new parts to the patterns.
These changes also mean that the website will look very different in a couple of months, so 3rd version arrives. The transition will also encompass new developer documentation (a transition guide of sorts) and extra effort to move from v2 to v3. Time frame: approximately 1st quarter of 2023.
The question is if the community is on board with that?

_Bobgeorgethe3rd’s feedback: what the transition between v2 and v3 encompasses. Answer: It won’t be a clear-cut transition, both will be maintained but v3 will be the default in the future.
The other question is what tools will be available for the users. Answer: It will be easy enough to use without any coding knowledge but also users will have the same capabilities as developers so they can use it if they wish to.
Are suggestions on UI/UX/layout invited? Answer: Yes, especially is you have strong opinions about it._

_Zwaluw’s feedback: Can people with more basic knowledge of JavaScript help somehow? Answer: Yes, new users are welcome to give feedback since they come with fresher viewpoint._

_Joost’s and Benjamin’s feedback: One of the ideas is to implement packs: for example, a sleeve pack and a collar pack. Under the hood, a pattern and a pack would both consist out of different parts but a pack would have only sleeves, only collars etc.
The consensus was reached that people are excited for FreeSewing v3._

✨ **Karen**

- Newsletter!

Every quarter there is a newsletter that includes all the news and updates, new patterns, sometimes interviews from FS contributors. The question is: are you interested about a certain contributor that you would like to know more about?
_Karen’s spotlight of people to interview: @SeaZeeZee, @eriese._
Remind Joost about an SPF record: issue #2616

✨ **Zwaluw**
Almost done with a new pattern for the FreeSewing! Bob can help with documentation. Should they wait until v3?
_Joost’s feedback: You don’t need to wait until v3 is released to post a new pattern since v3 will come in its time, you’re not tied to that timeline._

✨ **Bobgeorgethe3rd**
Name spaces and titles! It is not clear right now where parts need to go. Some titles on the website don’t use name spaces anymore. Maybe each pattern would need its own name space.
_Eriese’s feedback: we do need some guidance regarding name spaces.
Joost’s feedback: name spaces will change a bit in v3. Translations do need to be updated a bit. It needs some TLC._

✨ **Morgan**
Catherine right now cannot be drafted for people with profile measurements without breasts.
_Joost’s feedback: It is because it requires some measurements that are only presented in with breasts measurement profile. The measurements are planned to be changed in the future, have a bit of an update._

✨ **Nchilada**
It would be helpful to have more measurement in general to accommodate more types of bodies. Maybe blocks could have more shaping variants.

_Eriese’s feedback: it might be useful to have an option to put in the measurements people need/want to put in since some would only need some for time or comfort._

_Bob’s and Joost’s feedback: we are going towards more measurement and optional measurements, it will accommodate more bodies and will be useful to both beginners and advanced sewers._

_Joost: if people have a specific thing that they know works for them – please reach out, it might be useful to implement in into the website._

_Zwaluw’s feedback: It would be useful to add it to the newsletter so more people know that they can help by being a body ambassador (which is also a page on the website)._

_Bob’s feedback: There is a discussion on GitHub that can be useful for people to talk more about measurement adjustments and body shaping._

_**Karen’s feedback: I wonder if sometime before v3 release, it would be good to have a call item around "what do you wish the patterns accommodated for you/your body?"**_

**Comment by joostdecock** [comment]
Closing to increase visibility of active discussions
